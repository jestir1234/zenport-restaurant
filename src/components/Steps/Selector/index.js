import React from "react";
import PropTypes from "prop-types";
import Select from "react-select";
import "react-select/dist/react-select.css";
import styled, { keyframes } from "react-emotion";
import { MEDIA_BREAKPOINTS, COLORS } from "../../../constants";

const fade = keyframes`
  from { opacity: 0; }
  50%  { opacity: .5 }
  to   { opacity: 1 }`;

const Container = styled("div")(props => ({
  animation: `${fade} 1s ease-in-out`,
  padding: "20px 10px 0 10px",
  ...props.styles
}));

const Header = styled("h3")(props => ({
  whiteSpace: "nowrap",
  [MEDIA_BREAKPOINTS.down("xs")]: {
    whiteSpace: "normal"
  }
}));

const Error = styled("span")(props => ({
  color: COLORS.red,
  position: "absolute",
  maxWidth: "300px"
}));

const Selector = props => {
  return (
    <Container styles={props.styles}>
      <Header>{props.title}</Header>
      <Select
        name="form-field-name"
        value={props.selectedValue}
        onChange={props.handleChange}
        options={props.options}
      />
      <Error>{props.errorMessage}</Error>
    </Container>
  );
};

Selector.propTypes = {
  styles: PropTypes.object,
  title: PropTypes.string,
  value: PropTypes.string,
  handleChange: PropTypes.func,
  options: PropTypes.array
};

export default Selector;
