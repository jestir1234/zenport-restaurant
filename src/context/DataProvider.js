import React, { Component } from "react";
import data from "../data.json";

export const DataContext = React.createContext();

const parseRestaurants = () => {
  let restaurants = {};
  let uniqRestaurants = Array.from(
    new Set(data[0].dishes.map(dish => dish.restaurant))
  );
  uniqRestaurants.forEach(name => {
    restaurants[name] = { availableMeals: "", menu: [] };
  });

  data[0].dishes.forEach(dish => {
    let currentMenu = restaurants[dish.restaurant].menu;
    currentMenu.push(dish.name);
    restaurants[dish.restaurant] = {
      availableMeals: dish.availableMeals,
      menu: currentMenu
    };
  });
  return restaurants;
};

const parseMealTypes = restaurants => {
  let mealTypes = {};
  Object.keys(restaurants).forEach(restaurant => {
    let restaurantMealTypes = restaurants[restaurant]["availableMeals"];
    restaurantMealTypes.forEach(type => {
      if (mealTypes[type]) {
        mealTypes[type]["restaurants"][restaurant] = restaurants[restaurant];
      } else {
        mealTypes[type] = { restaurants: {} };
      }
    });
  });
  return mealTypes;
};

const defaultStepOptions = () => ({
  step1: {
    selectedMealType: null,
    selectedPartyCount: { value: 1, label: 1 }
  },
  step2: {
    selectedRestaurant: null
  },
  step3: {
    selectedMeal: null,
    selectedServings: { value: 1, label: 1 },
    currentOrder: {},
    validMealCount: false,
    errorMessage: ""
  }
});

class DataProvider extends Component {
  state = {
    mealTypes: parseMealTypes(parseRestaurants()),
    stepOptions: defaultStepOptions()
  };

  handleChange = (selectedOption, option, currentStep) => {
    this.setState(prevState => ({
      ...prevState,
      stepOptions: {
        ...prevState.stepOptions,
        [currentStep]: {
          ...prevState.stepOptions[currentStep],
          [option]: selectedOption
        }
      }
    }));
  };

  addOrder = order => {
    let newOrder = { ...this.state.stepOptions.step3.currentOrder, ...order };
    this.setState(
      prevState => ({
        ...prevState,
        stepOptions: {
          ...prevState.stepOptions,
          step3: {
            ...prevState.stepOptions.step3,
            currentOrder: newOrder
          }
        }
      }),
      this.validateMealCount
    );
  };

  subtractOrder = dish => {
    let currentOrder = { ...this.state.stepOptions.step3.currentOrder };
    delete currentOrder[dish];
    this.setState(
      prevState => ({
        ...prevState,
        stepOptions: {
          ...prevState.stepOptions,
          step3: {
            ...prevState.stepOptions.step3,
            currentOrder
          }
        }
      }),
      this.validateMealCount
    );
  };

  validateMealCount = () => {
    let currentOrders = this.state.stepOptions.step3.currentOrder;
    let totalMeals =
      Object.keys(currentOrders).length > 0
        ? Object.values(this.state.stepOptions.step3.currentOrder).reduce(
            (acc, val) => acc + val
          )
        : 0;

    let selectedPartyCount = this.state.stepOptions.step1.selectedPartyCount
      .value;
    let errorMessage = "";

    if (totalMeals > 10) {
      errorMessage = "Exceeded max total servings of 10.";
    }

    if (totalMeals < selectedPartyCount) {
      errorMessage =
        "Total servings must be equal to or greater than the no. of people in your group.";
    }

    this.setState(prevState => ({
      ...prevState,
      stepOptions: {
        ...prevState.stepOptions,
        step3: {
          ...prevState.stepOptions.step3,
          validMealCount: totalMeals <= 10 && totalMeals >= selectedPartyCount,
          errorMessage
        }
      }
    }));
  };

  render() {
    return (
      <DataContext.Provider
        value={{
          state: this.state,
          handleChange: (selectedOption, option, currentStep) =>
            this.handleChange(selectedOption, option, currentStep),
          addOrder: order => this.addOrder(order),
          subtractOrder: dish => this.subtractOrder(dish)
        }}
      >
        {this.props.children}
      </DataContext.Provider>
    );
  }
}

export default DataProvider;
