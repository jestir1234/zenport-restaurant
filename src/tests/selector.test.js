import React from "react";
import Selector from "../components/Steps/Selector";
import renderer from "react-test-renderer";
import DataProvider from "../context/DataProvider";
const PARTY_MAX = 10;

const context = {
  handleChange: (e, selectedPartyCount, step) =>
    console.log("handleChangeFunction"),
  state: {
    stepOptions: {
      step1: {
        selectedPartyCount: 4
      }
    }
  }
};

test("Renders the <Selector /> component", () => {
  const component = renderer.create(
    <Selector
      title={"Enter the number of people in your party"}
      options={[...Array(PARTY_MAX)].map((count, idx) => {
        return {
          value: idx + 1,
          label: idx + 1
        };
      })}
      handleChange={e =>
        context.handleChange(e, "selectedPartyCount", this.state.step)
      }
      selectedValue={context.state.stepOptions["step1"]["selectedPartyCount"]}
    />
  );
  let tree = component.toJSON();
  expect(tree);

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
