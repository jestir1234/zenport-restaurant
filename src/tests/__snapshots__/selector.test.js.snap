// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the <Selector /> component 1`] = `
<div
  className="css-1uh653v"
>
  <h3
    className="css-v9loj0"
  >
    Enter the number of people in your party
  </h3>
  <div
    className="Select has-value is-clearable is-searchable Select--single"
    style={undefined}
  >
    <input
      disabled={false}
      name="form-field-name"
      type="hidden"
      value="4"
    />
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={undefined}
    >
      <span
        className="Select-multi-value-wrapper"
        id="react-select-2--value"
      >
        <div
          className="Select-value"
          style={undefined}
          title={undefined}
        >
          <span
            aria-selected="true"
            className="Select-value-label"
            id="react-select-2--value-item"
            role="option"
          >
            4
          </span>
        </div>
        <div
          className="Select-input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-activedescendant="react-select-2--value"
            aria-describedby={undefined}
            aria-expanded="false"
            aria-haspopup="false"
            aria-label={undefined}
            aria-labelledby={undefined}
            aria-owns=""
            className={undefined}
            id={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "5px",
              }
            }
            tabIndex={undefined}
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </span>
      <span
        aria-label="Clear value"
        className="Select-clear-zone"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        title="Clear value"
      >
        <span
          className="Select-clear"
          dangerouslySetInnerHTML={
            Object {
              "__html": "&times;",
            }
          }
        />
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
  <span
    className="css-rwbn3g"
  />
</div>
`;
